### DISCUSSION TITLE
What is the best way to publish winui 3 app for the store?

### DISCUSSION THREAD

**Discussion by ghost1372** [microsoft/WindowsAppSDK#3923] [Q&A] [answered]
What is the best way to publish winui 3 app for the store?
Currently I am publishing with the following settings:
```
WindowsAppSDKSelfContained = true
PublishReadyToRun = true

```
my generated MSIX file is almost 80 mb, which is a bit much, and if I output for all three platforms the total is 200MB
And it takes a little time for me to upload it. 

Is there a better way to publish? If I disable the WindowsAppSDKSelfContained, will the app be installed and run from the store without any problems?

**Comment by DarranRowe** [comment, marked as answer]
While it doesn't help having WindowsAppSDKSelfContained set to true, the real problem here is .NET.
Suppose I create a new C# project using the C# version of the "Blank App, Packaged (WinUI 3 in Desktop)" template. Update all of the NuGet packages to the latest version. I then use the Package & Publish option to create the packages. Each individual .msix package will be around 48MiB.
Suppose I follow the same steps using a C++ project using the C++ version of the "Blank App, Packaged (WinUI 3 in Desktop)" template. Each individual .msix package will be around 150KiB.
If you look at the contents of the packages, you will notice a rather substantial difference.

![Screenshot 2023-10-26 140008](https://github.com/microsoft/WindowsAppSDK/assets/52577874/36262b4c-5b18-468d-bbc7-13c6066e6d4c)

![Screenshot 2023-10-26 135914](https://github.com/microsoft/WindowsAppSDK/assets/52577874/d11510c7-609d-4860-baec-61e5322335bb)

The C# package contains so much more compared to the C++ package, and some of those file in the C# package should look familiar. First is the C# projection, this takes up 8MiB. Second:

![Screenshot 2023-10-26 140555](https://github.com/microsoft/WindowsAppSDK/assets/52577874/45d26c78-bf75-4a30-bf2d-efb1984ddf75)

Yes, that is the .NET runtime. If you look at the publish profile, the C# projects are publishing self contained.
So, the glib and unhelpful response to your question is that the best way to publish WinUI 3 apps is using the C++ projects.
To provide a more helpful response, there is nothing that you can do yet. I can't find any reference to there being a .NET framework package and if you try to disable .NET self contained then it will require the end user to install the correct .NET version.
One technology that has been around in a limited form since .NET 7 is Native AOT. The .NET 7 version was only targeted at console applications though. The version in .NET 8 is more comprehensive and is targeted at UI applications, and, when supported, will be the preferred way of publishing applications. Since this will be compiling .NET code to native, which includes IL trimming, then the code will end up much smaller and will not require including the .NET runtime with the package.
So basically, you will have to wait until Native AOT is supported in the Windows App SDK/WinUI 3. You don't have to make the Windows App SDK self contained though, it will add a package reference to the AppxManifest.xml file and the Windows App Runtime will be installed along side your app when you install it from the store.
